File: Problems/Algorithms/15. 3Sum/3Sum.kt
class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        var n = nums.size
        if (n <= 2) return mutableListOf()
        
        nums.sort()
        
        val results = mutableListOf<List<Int>>()
        for (i in 0..n-2) {
            if (i > 0 && nums[i] == nums[i-1]) {
                continue
            }
            
            var j = i + 1
            var k = n - 1
            while (j < k) {
                var sum = nums[i] + nums[j] + nums[k]
                if (sum == 0) {
                    results.add(listOf(nums[i], nums[j], nums[k]))
                    
                    k--
                    while (j < k && nums[k] == nums[k+1]) {
                        k--
                    }
                } else if (sum > 0) {
                    k--
                } else {
                    j++
                }
            }
        }
        
        return results
    }
}

File: Problems/Algorithms/15. 3Sum/3_sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if n <= 2:
            return []

        nums.sort()

        results = []
        for i in range(n-1):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            j, k = i+1, n-1
            while j < k:
                total = nums[i] + nums[j] + nums[k]
                if total == 0:
                    results.append([nums[i], nums[j], nums[k]])

                    k -= 1
                    while j < k and nums[k] == nums[k+1]:
                        k -= 1
                elif total > 0:
                    k -= 1
                else:
                    j += 1

        return results

